consensus/src/detail/mod.rs

```diff
@@ -200,6 +208,119 @@ impl Consensus {
+        height: &BlockHeight,
+    ) -> Result<Option<BlockHeader>, BlockError> {
+        let consensus_ref = self.make_ro_db_tx();
+        let id = consensus_ref
+            .db_tx
+            .get_block_id_by_height(height)?
+            .ok_or(BlockError::NotFound)?;
+        consensus_ref.db_tx.get_block_header_by_id(&id).map_err(BlockError::from)
+    }
+
+    pub fn get_locator(&self) -> Result<Vec<BlockHeader>, BlockError> {
+        let id = self.get_best_block_id()?.ok_or(BlockError::NotFound)?;
+        let height = self.get_block_height_in_main_chain(&id)?.ok_or(BlockError::NotFound)?;
+
+        let headers = itertools::iterate(0, |&i| if i == 0 { 1 } else { i * 2 })
+            .take_while(|i| (height - BlockDistance::new(*i)).is_some())
+            .map(|i| {
+                self.get_header_from_height(
+                    &(height - BlockDistance::new(i)).expect("distance to be valid"),
+                )
+            })
+            .filter(|res| res.is_ok())
+            .filter_map(|res| res.unwrap())
+            .collect::<Vec<_>>();
+
+        Ok(headers)
+    }
+
+    // TODO: use `ConsensusRef::is_block_in_main_chain()` implementation instead, how?
+    fn is_block_in_main_chain(&self, block_index: &BlockIndex) -> Result<bool, BlockError> {
+        let consensus_ref = self.make_ro_db_tx();
+        let height = block_index.get_block_height();
+        let id_at_height =
+            consensus_ref.db_tx.get_block_id_by_height(&height).map_err(BlockError::from)?;
+        match id_at_height {
+            Some(id) => Ok(id == *block_index.get_block_id()),
+            None => Ok(false),
+        }
+    }
+
+    pub fn get_headers(&self, locator: Vec<BlockHeader>) -> Result<Vec<BlockHeader>, BlockError> {
+        // use genesis block if no common ancestor with better block height is found
+        let mut best = BlockHeight::new(0);
+
+        for header in locator.iter() {
+            if let Some(block_index) = self.get_block_index(&header.get_id())? {
+                if self.is_block_in_main_chain(&block_index)? {
+                    best = block_index.get_block_height();
+                    break;
+                }
+            }
+        }
+
+        // get headers until either the best block or header limit is reached
+        let limit = std::cmp::min(
+            (best + HEADER_LIMIT).ok_or(BlockError::Unknown)?,
+            self.get_block_height_in_main_chain(
+                &self.get_best_block_id()?.expect("best block to exist"),
+            )?
+            .expect("best block's height to exist"),
+        );
+
+        let headers = itertools::iterate(best.next_height(), |iter| iter.next_height())
+            .take_while(|height| height <= &limit)
+            .map(|height| self.get_header_from_height(&height))
+            .filter(|res| res.is_ok())
+            .filter_map(|res| res.unwrap())
+            .collect::<Vec<_>>();
+
+        Ok(headers)
+    }
+
+    pub fn get_uniq_headers(
```

Comment:
Having looked at this function again, I'd argue it is useful and worth keeping. The reason is because the locator object is inherently inaccurate which means that the response we receive is most likely going to contain headers we already know. For example, if we are on chain `a - b - c - d - e - f - g - h - i - j - k - l - m`, our locator is  `m - l - k - i - e`.

The node we connect to is tracking chain `a - b - c - d - e - f - g -  h - N1 - N2 - N3`, but because the locator they received attached to their chain at `e`, they send us `f - g -  h - N1 - N2 - N3` which attaches to our chain but contains blocks that we already know.

What this function did was verify that the headers we received in response to `GetHeaders` attached to our chain and that they were in order. While verifying the order, it would select all headers we don't know about and report those to syncing.

I'd argue this function is useful and we should have it in the consensus interface.

Reply:
The problem is that this function is not written correctly in a way that solves the minimal problem it's meant to be solving (the problem you described in this comment). This is the primary issue. So, I'm gonna make a few suggestions on how to make this function useful and minimal.

1. The function name should be `filter_already_existing_blocks()`.
2. The last loop is mostly pointless and redundant. The chainstate cannot have any blocks that don't have a previous block in the blockchain, which means, you should just break once the block is not found.
3. You should check that blocks are a "chain" in P2P, not in consensus. No reason to delegate this call.

Meaning, this whole function could be changed to the following:
1. Check that we have the previous block at headers[0] (please remove the check for empty, and then use get(0) which will incorporate that
2. Loop over all blocks and check that they exist. Once you hit a block we don't know, break the loop, erase everything from the beginning to that point, and return the new headers.
3. In P2P, before calling this function, make sure that all headers are consecutive. This property can be a `debug_assert!()` in this function.